source/fe/fe_simplex_p.cc

```diff
@@ -436,6 +438,138 @@ FE_SimplexPoly<dim, spacedim>::get_prolongation_matrix(
+      // along with the number of the DoF on this vertex
+      const unsigned int face_vertex = face_index / this->n_dofs_per_vertex();
+      const unsigned int dof_index_on_vertex =
+        face_index % this->n_dofs_per_vertex();
+
+      // then get the number of this vertex on the cell and translate
+      // this to a DoF number on the cell
+      return this->reference_cell().face_to_cell_vertices(
+               face, face_vertex, combined_orientation) *
+               this->n_dofs_per_vertex() +
+             dof_index_on_vertex;
+    }
+  else if (face_index < this->get_first_face_quad_index(face))
+    // DoF is on a line
+    {
+      // do the same kind of translation as before. we need to only consider
+      // DoFs on the lines, i.e., ignoring those on the vertices
+      const unsigned int index =
+        face_index - this->get_first_face_line_index(face);
+
+      const unsigned int face_line         = index / this->n_dofs_per_line();
+      const unsigned int dof_index_on_line = index % this->n_dofs_per_line();
+
+      // we now also need to adjust the line index for the case of
+      // face orientation, face flips, etc
+      unsigned int adjusted_dof_index_on_line = 0;
+      switch (dim)
+        {
+          case 1:
+            DEAL_II_ASSERT_UNREACHABLE();
+            break;
+
+          case 2:
+            if (combined_orientation == numbers::default_geometric_orientation)
+              adjusted_dof_index_on_line = dof_index_on_line;
+            else
+              adjusted_dof_index_on_line =
+                this->n_dofs_per_line() - dof_index_on_line - 1;
+            break;
+
+          case 3:
+            // in 3d, things are difficult. someone will have to think
+            // about how this code here should look like, by drawing a bunch
+            // of pictures of how all the faces can look like with the various
+            // flips and rotations.
+            //
+            // that said, we can implement a couple other situations easily:
+            // if the face orientation is
+            // numbers::default_combined_face_orientation then things are
+            // simple. Likewise if the face orientation is
+            // numbers::reverse_line_orientation then we just know we need to
+            // reverse the DoF order on each line.
+            //
+            // this at least allows for tetrahedral meshes where periodic face
+            // pairs are each in one of the two above configurations. Doing so
+            // may turn out to be always possible or it may not, but at least
+            // this is less restrictive.
+            Assert((this->n_dofs_per_line() <= 1) ||
+                     combined_orientation ==
+                       numbers::default_geometric_orientation ||
+                     combined_orientation == numbers::reverse_line_orientation,
+                   ExcNotImplemented());
+
```

Comment:
```suggestion
            Assert((this->n_dofs_per_line() <= 1) ||
                     combined_orientation ==
                       numbers::default_geometric_orientation,
                   ExcNotImplemented());

```
Do the tests still pass if we only support things in the default configuration? The test you added has a pretty basic periodicity setup.

Reply:
Unfortunately not. In fact, I looked carefully, but not exhaustively, for a way to split a hex into 6 tetrahedra satisfying:

1) Each cell has positive measure.
2) Each external face is in the default orientation.

After looking at this for a long time, I doubt such a configuration exists. This is the reason I extended to more than one possible orientation.

In fact my test case ends up being such that each combined face orientation is the "mirror" state, so it's interesting that it even works given your comments above.

If it is possible in dealii to mark tetrahedra cells as having negative measure, then this is an easy fix. Otherwise I will have to dig into the complexities you mentioned in another comment.

Reply:
Ok, I had a quick look at my test case and it seems that indeed all face pairs have combined orientation "0". Only five of the possible 10 combinations of local face indices occur: (0,0), (0,2), (1,2), (1,3), (2,3).

So then, if I understand you right, you are saying my code will break for one of the remaining combinations of local face indices. Is that correct?

Reply:
Sorry, I didn't want to address the other changes until we have a chance to iron this particular thread out--since it kind of seems to me this concern underlies the validity of the whole pull request.

If you have a chance to confirm that my comments capture the cases you're concerned about or clarify where you believe the problem could arise, I'm happy to add a test to this pull request which covers them and adjust this function as needed.